### ISSUE TITLE
Backport of security patch, for benefit of yargs

### ISSUE THREAD

**Issue by bcoe** [chalk/ansi-regex#38]
I know it's a pain in the neck, but would you consider back-porting https://github.com/chalk/ansi-regex/pull/37 to the 5.x.x release line, for the benefit of [yargs](https://github.com/yargs/yargs/issues/1839#issuecomment-918380545).

Yargs is making the effort during the transition to ESM to support both CJS and ESM, which makes us unable to update to the latest version of `string-width`.

If you were willing to make an exception (_I know you're pushing folks towards using ESM exclusively_) it would be really valuable for yargs users using CJK character sets.

**Comment by Qix-** [comment]
Sure thing, give me a moment.

**Comment by Qix-** [comment]
Released as `5.0.1` :) Let me know if there's anything else.

**Comment by bacali95** [comment]
Hello @yetingli, can you please edit the vulnerability information in Snyk (https://app.snyk.io/vuln/SNYK-JS-ANSIREGEX-1583908) since the vulnerability fix is backported to version 5.0.1? Thanks :smile:

**Comment by yetingli** [comment]
Hey @bacali95, I don't seem to have permission to edit them. Hi Colin @colin-ife-snyk, can you please edit the vulnerability information? Thanks a lot!

**Comment by colin-ife-snyk** [comment]
> Hey @bacali95, I don't seem to have permission to edit them. Hi Colin @colin-ife-snyk, can you please edit the vulnerability information? Thanks a lot!

No problem. Will be updated later today.

**Comment by bacali95** [comment]
Thanks a lot @yetingli @colin-ife-snyk :rose:

**Comment by sindresorhus** [comment]
The range in the report says `<5.0.1` but it has not been proven that all versions below `5.0.0` are vulnerable.

**Comment by sindresorhus** [comment]
@colin-ife-snyk @yetingli ⬆️

**Comment by cardil** [comment]
GitHub Advisory Database still lists this package as affected for `<6.0.1`, here: https://github.com/advisories/GHSA-93q8-gq69-wqmw

Anyone knows how to update that information as well?

This page https://docs.github.com/en/code-security/security-advisories/creating-a-security-advisory suggest the repo owner should be able to edit this information.

**Comment by cardil** [comment]
I've raised the affected field on Github Securitylab discussion as well: https://github.com/github/securitylab/discussions/436

**Comment by bcoe** [comment]
@Qix- thank you very much for the back port 😄, I know this can be a pain in the neck to do.

**Comment by Qix-** [comment]
No problem at all, glad I could help :)

**Comment by sindresorhus** [comment]
I have confirmed now that this vulnerability exists in v4.1.0, v4.0.0, v3.0.0, but **not** v2.1.1 and lower.

**Comment by sindresorhus** [comment]
@yetingli For future reports, please include the affected version range, and a regression test if you decide to submit a fix. Thanks :)

**Comment by sindresorhus** [comment]
@colin-ife-snyk Can you please update the Snyk report? :arrow_up:

**Comment by yetingli** [comment]
@sindresorhus Thanks for reminding.  I will pay attention to this in the future.

**Comment by benjifin** [comment]
@sindresorhus updated as well now in our (Snyk's) backend - will take a few hours to propagate to prod. Thanks!

**Comment by G-Rath** [comment]
@cardil cheers, that's a good workaround for the short-term but I'd like to avoid it if possible since this affects pretty much all of our projects - while I said `serverless` in my previous comment, really it also affects PHP & Ruby projects too, as they've got `laravel-mix` & `@rails/webpacker` so they both end up pulling in versions of big packages like `webpack-dev-server` & `webpack-cli` that in turn are pulling in versions of `yargs` that end up using v3 & v4 of `ansi-regex`, which we can't do much about because of the nature of these packages as being monolithic framework/"we'll handle the js build side for you" type cli tools.

This is also why I expect it'll take 2-3 years for them to eventually move off the packages that require the v3 & v4 versions of `ansi-regex`, because they don't tend to release new majors frequently because of that nature and that'll also mean the upgrades will likely require some elbow grease (and assuming all our clients would be happy to sign off on the work).

Getting patch releases for these other versions would mean we can avoid having to perpetually ignore these vulnerabilities (or `preinstall` patch them) for pretty much every new & existing project for those years (and not have to spend time justifying them to pentesters, govt heads-of-security, etc).

**Comment by Qix-** [comment]
Probably not what you want to hear/read, but I've written on backports at-length here: https://github.com/yargs/yargs/issues/1839#issuecomment-942572494

**Comment by ljharb** [comment]
The sad part is that while I agree with you on the problem, your stance won't actually change the perverse incentives in the ecosystem - it will just cause harm for users.

**Comment by G-Rath** [comment]
@Qix- I do generally agree with your view on this, but unfortunately it's not something that is always with our control (which greatly frustrates me too, and I know will frustrate you to hear) - while I've spent the last two weeks looking into how we can get things upgraded (including updating libraries to be using higher versions of `ansi-regex` where possible), I think I have reached the limit that I can do so.

One of the chief packages that we cannot upgrade is `webpack-dev-server` - we're locked into v3 by [`@rails/webpacker`](https://github.com/rails/rails/issues/43062), which means we're stuck waiting for them to release v6 (which uses v4 of wds); this is complicated by the fact that the [Rails team is currently planning to move _off_ `@rails/webpacker`](https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755), leaving its future unclear and the likelihood of v6 being released anytime soon unlikely.

I plan to look into seeing if I can implement support for v4 of wds into v5 of `@rails/webpacker`, but I suspect that even if I could it would be unlikely to get reviewed, let alone merged and released.

This leaves us in a difficult spot having to deal with this, because while I'd love to do things like try and chuck out `@rails/webpacker` , that would not be feasible for us to do since it would take a huge amount of time (since it'd been to be done for all of our Ruby on Rails applications, of which there are many) and risk (since we'd need to be rolling our own solution, or try and leverage the currently-experimental replacement that could change under us).

This is a situation that frustrates me greatly, because I agree it's not your fault nor your burden to bear, and those who should be bearing it are not; likewise for me, I have to explain to clients and heads-of-securities about how "no it's not actually that big of a risk, it's 'high' because in theory but in reality... etc".

This all leads me back to here, asking if it would be possible to get just a v4 bugfix applying the security fix:

```
diff --git a/index.js b/index.js
index c254480..9e37ec3 100644
--- a/index.js
+++ b/index.js
@@ -6,7 +6,7 @@ module.exports = options => {
        }, options);

        const pattern = [
-               '[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)',
+               '[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]+)*|[a-zA-Z\\d]+(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)',
                '(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PR-TZcf-ntqry=><~]))'
        ].join('|');
```

Based on my efforts to get libraries upgraded, I've found that the ones I couldn't are typically because they still support Node v6 meaning they're limited to using v4 of `ansi-regex` - I can workaround the higher-level libraries like `has-ansi` not having a release that supports both Node v6 _and_ using `ansi-regex` v4 easily enough with inlining, but only if there's a "secure" v4 of `ansi-regex`.
